### PR TITLE
968 create error page

### LINKED ISSUE #968: Create an Error Page
We should handle errors more cleanly than allowing the default error page to show to the user.

One error that pops to mind is the dynamic module issue (when we update the site, but users still have an older version of the site open).

Screenshot 2024-06-03 at 2 40 56 PM

### PR DESCRIPTION
**Pull request by alexespejo**
## Summary
Created a custom error page that displays when a user encounters an error or enters a bad route 

image
image

## Issues
In dark mode the link text color poorly contrasts the gray background 

## Test plan
This error page should appear when the user:
- Enters an invalid route (ex. /test/test)
- A runtime error is caught 
   - To test for runtime errors a `throw Error ('some message')` on any of the pages to simulate an error being thrown
- When an update is made to prod while the user is using the app (in theory)

## Issues
- Closes #968

## Future Follow-Up
- #1047

### CODE REVIEW

File: apps/antalmanac/src/routes/Error.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/routes/Error.tsx]:
```suggestion
                This error may be caused by your browser having an out of date version of AntAlmanac.
```
nit: "you" may feel too direct for users

File: apps/antalmanac/src/routes/Error.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/routes/Error.tsx]:
```suggestion
            Oops! Something went wrong.
```
nit: 🥺 as much as I'd like to keep the emojis, I think AA broadly tries to maintain a professional tone / communication

File: apps/antalmanac/src/routes/Error.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/routes/Error.tsx]:
```suggestion
```
nit: prefer non-default exports (see above)

File: apps/antalmanac/src/routes/Error.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/routes/Error.tsx]:
```suggestion
export const ErrorPage = () => (
```
nit: prefer exporting non-default components. Although large portions of the AA codebase uses default exports, it's better practice to export (and import) components more explicitly

**Review by KevinWu098 — CHANGES_REQUESTED** [review]
Congrats on your first PR! Mostly looks good to me, with some nitpicks on wording and best practices.

Two main points other than the review comments:
One: If you're feeling up to it, I'd say a centered (but still mobile-friendly) error page would be better suited for making the error page look cleaner (see below)

image

No need to add graphics and whatnot, but centering does make the error page (IMO) feel nicer.


Two: Could you include screenshots of the error page (both light and dark) in your pull request? (it helps your reviewer visually examine the page, although we usually checkout the branch locally as well).

**Review by MinhxNguyen7 — CHANGES_REQUESTED** [review]
Also, can you add reproduction information in the description so that reviewer(s) don't have to figure out a way to induce an error? That should be under a section titled "Test Plan". See [this PR](https://github.com/icssc/AntAlmanac/pull/1008) as an example for writing a PR description
Also, please specify which issue it closes there? I know it's in the title, but putting it in the description closes it automatically when the PR merges

**Comment by alexespejo** [comment]
Edited message (dark and light) 

image
image

**Comment by alexespejo** [comment]
> Also, can you add reproduction information in the description so that reviewer(s) don't have to figure out a way to induce an error? That should be under a section titled "Test Plan". See [this PR](https://github.com/icssc/AntAlmanac/pull/1008) as an example for writing a PR description Also, please specify which issue it closes there? I know it's in the title, but putting it in the description closes it automatically when the PR merges

Is there a known way to reproduce runtime errors like a bad fetch request. In theory the page should appear for all errors but I wanted to see

File: apps/antalmanac/src/routes/Error.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/routes/Error.tsx:1]: Also noting that, if possible, we should also display the actual error that we're catching to the user (so that they can then pass it along to us via screenshot or copy/paste).

**Comment by KevinWu098** [comment]
> > Also, can you add reproduction information in the description so that reviewer(s) don't have to figure out a way to induce an error? That should be under a section titled "Test Plan". See [this PR](https://github.com/icssc/AntAlmanac/pull/1008) as an example for writing a PR description Also, please specify which issue it closes there? I know it's in the title, but putting it in the description closes it automatically when the PR merges
> 
> Is there a known way to reproduce runtime errors like a bad fetch request. In theory the page should appear for all errors but I wanted to see

mm, might be a StackOverflow question -- one thing that jumps to mind is just writing in a `throw Error` into any given file!

**Comment by MinhxNguyen7** [comment]
> Is there a known way to reproduce runtime errors like a bad fetch request. In theory the page should appear for all errors but I wanted to see

Ah I didn't realize you hadn't figured that.

**Comment by alexespejo** [comment]
> > > Also, can you add reproduction information in the description so that reviewer(s) don't have to figure out a way to induce an error? That should be under a section titled "Test Plan". See [this PR](https://github.com/icssc/AntAlmanac/pull/1008) as an example for writing a PR description Also, please specify which issue it closes there? I know it's in the title, but putting it in the description closes it automatically when the PR merges
> > 
> > 
> > Is there a known way to reproduce runtime errors like a bad fetch request. In theory the page should appear for all errors but I wanted to see
> 
> mm, might be a StackOverflow question -- one thing that jumps to mind is just writing in a `throw Error` into any given file!

this works

**Comment by alexespejo** [comment]
New error page displaying the error message. 

The error message can be hidden by default if anyone has a preference

Toggle on 
image
image


Toggle off 
image

File: apps/antalmanac/src/routes/Error.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/routes/Error.tsx]:
```suggestion
import { Link, useRouteError } from 'react-router-dom';
```
nit: no dupe imports

File: apps/antalmanac/src/routes/Error.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/routes/Error.tsx]: nit: name your file the same as the component (or vice versa)

File: apps/antalmanac/src/routes/Error.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/routes/Error.tsx]: Is the comment intended? Looks like it's always open?

Regardless, I think we should keep the error message always displayed, as we want users to immediately (and visibly) see the error so they're more likely to pass it along to us.

**Review by KevinWu098 — CHANGES_REQUESTED** [review]
Looks awesome! Really solid improvement, and will hopefully help our bug reporting user flow.

Could you update your original PR message to
1. Include the latest images
2. Update the `Issues` section to link to the original issue #968 (e.g. `Closes #968`)
    - the issues section isn't really used to state issues about your PR, but rather to just write the issues this PR is related to / closes

Just a couple more nits (I swear I'm not always this pedantic), but otherwise looks solid! Will also ping @MinhxNguyen7 again for another pair of eyes, since this page will be relatively prominent to users.

File: apps/antalmanac/src/routes/Error.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/routes/Error.tsx]:
```suggestion
                    <Link to="https://forms.gle/k81f2aNdpdQYeKK8A">bug report</Link> with the provided error.
```
nit: Missing a period :p

**Review by KevinWu098 — APPROVED** [review]
LGTM, but I'll hold off on merging until Minh or Dalton gets a chance to look at this

**Review by MinhxNguyen7 — APPROVED** [review]

**Pull request merged**
